### PR TITLE
ARROW-14602: [Doc] Tutorial - Python feature PR

### PR DESCRIPTION
**Pull request by AlenkaF**
This is a draft of the **Python Tutorial** that will be included in the (New) Contributors Guide.

It is a tutorial explaining the steps in making [ARROW-14977: [Python] Add a "made-up" feature for the guide tutorial](https://github.com/apache/arrow/pull/11900).

### CODE REVIEW

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 1]
```diff
@@ -25,3 +25,513 @@
...
+   INFO:archery:Running Python formatter (autopep8)
+   INFO:archery:Running Python linter (flake8)
+
+Done. Now lets make the Pull Request!
+
+Creating a Pull Request
+-----------------------
+
+First lets check again the changes made to the local copy
+of the Arrow library in the shell.
```
[the remark this answers is not in the document]
Comment (wjones127) [reply docs/source/developers/guide/tutorials/python_tutorial.rst:381]: Do you also want to mention why we do this step?

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 2]
```diff
@@ -25,3 +25,513 @@
...
+to the branch history):
+
+.. code:: console
+
+   $ git commit -am "Adding a new compute feature for tutorial purposes"
+   [ARROW-14977 170ef85be] Adding a new compute feature for tutorial purposes
+    2 files changed, 51 insertions(+)
+   
+
+check the history of commits
```
Comment (wjones127) [inline docs/source/developers/guide/tutorials/python_tutorial.rst:446]:
```suggestion
You can use `git log` to check the history of commits:
```

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 3]
```diff
@@ -25,3 +25,513 @@
...
+       We can use this in release process instead of building on release
+       manager's local environment.
+       
+       Closes #11856 from kou/ci-docs-upload
+       
+       Authored-by: Sutou Kouhei <kou@clear-code.com>
+       Signed-off-by: Sutou Kouhei <kou@clear-code.com>
+   ...
+
+if necessary we should rebase to upstream master:
```
Comment (wjones127) [inline docs/source/developers/guide/tutorials/python_tutorial.rst:472]: We should probably say when necessary:

```suggestion
If you started the branch some time ago, you may need to rebase to upstream master to make sure there are no merge conflicts:
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
Thank you for this very comprehensive guide! I've left a few suggestions.

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 4]
```diff
@@ -25,3 +25,513 @@
...
+section. Navigate there whenever there is some information
+you may find is missing here.
+
+The feature contribution will be added to the compute module
+in PyArrow. But you can also follow the steps in case you are
+correcting a bug or adding a binding.
+
+This tutorial is different from the :ref:`step_by_step` as we
+will be working on a specific case. This tutorial is not meant
+as a step by step guide.
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply docs/source/developers/guide/tutorials/python_tutorial.rst:41]: That is correct. I do not think Arrow is a project where ppl will go and learn things in this way but I can make it more explicit if needed. Step by step guide is added in the New Contributors Guide section and this Tutorial is more of a practical view of the contributing process.

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 5]
```diff
@@ -25,3 +25,513 @@
...
+
+.. code:: console
+   
+   $ archery lint --python --fix
+   INFO:archery:Running Python formatter (autopep8)
+   INFO:archery:Running Python linter (flake8)
+   /Users/alenkafrim/repos/arrow/python/pyarrow/tests/test_compute.py:2288:80: E501 line too long (88 > 79 characters)
+
+We get a warning that one line is too long. Lets
+make the necessary corrections and run it again.
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply docs/source/developers/guide/tutorials/python_tutorial.rst:367]: I will have to add a section on styling to the guide and will also add this info there.

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 6]
```diff
@@ -25,3 +25,513 @@
...
+
+Once the `Apache Arrow repository <https://github.com/apache/arrow>`_
+is forked we will clone it and add the link of the main repository
+to our upstream.
+
+.. code:: console
+
+   $ git clone https://github.com/<your username>/arrow.git
+   $ cd arrow
+   $ git remote add upstream https://github.com/apache/arrow
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply docs/source/developers/guide/tutorials/python_tutorial.rst:59]: That is right. This two ways of working are both valid I think. But would keep it as it is so it is in line with the docs written if you agree?

[the remark this answers is not in the document]
Comment (wjones127) [reply docs/source/developers/guide/tutorials/python_tutorial.rst:59]: Yeah feel free to do that however you want it 😄  Just wanted to share that.

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 7]
```diff
@@ -25,3 +25,532 @@
...
+This tutorial is different from the :ref:`step_by_step` as we
+will be working on a specific case. This tutorial is not meant
+as a step by step guide.
+
+**Lets start!**
+
+Set up
+------
+
+Lets setup the Arrow repository. We presume here that Git is
```
Comment (wjones127) [inline docs/source/developers/guide/tutorials/python_tutorial.rst:48]:
```suggestion
Let's setup the Arrow repository. We presume here that Git is
```

File: docs/source/developers/guide/tutorials/python_tutorial.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (wjones127) [reply docs/source/developers/guide/tutorials/python_tutorial.rst:41]:
```suggestion
will be working on a specific case. This tutorial is not meant
as a step-by-step guide.
```

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 8]
```diff
@@ -25,3 +25,532 @@
...
+
+   Adding a comment to the JIRA ticket we created.
+
+We have successfully created a new JIRA issue with index ARROW-14977.
+
+.. figure:: python_tutorial_jira_issue.jpeg
+   :scale: 50 %
+   :alt: JIRA page of the issue just created.
+
+   Our JIRA issue. Yay!
```
Comment (wjones127) [inline docs/source/developers/guide/tutorials/python_tutorial.rst:119]: This is done in the screenshot, but I think worth calling out to the reader: "Make sure you assign yourself to the issue to let others know you are working on it."

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 9]
```diff
@@ -25,3 +25,532 @@
...
+
+   And search through the ``test_compute.py`` file in ``pyarrow``
+   folder.
+
+From the search we can see that the function is tested in the
+``python/pyarrow/tests/test_compute.py`` file that would mean the
+function is defined in the ``compute.py`` file.
+
+After examining the ``compute.py`` file we can see that together
+with ``_compute.py`` the functions from C++ get wrapped into Python.
```
Comment (wjones127) [inline docs/source/developers/guide/tutorials/python_tutorial.rst:167]: It's either pyx or pyd, right?
```suggestion
with ``_compute.pyx`` the functions from C++ get wrapped into Python.
```

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 10]
```diff
@@ -25,3 +25,532 @@
...
+.. code:: console
+
+   $ cd python
+   $ python
+
+   Python 3.9.7 (default, Oct 22 2021, 13:24:00) 
+   [Clang 13.0.0 (clang-1300.0.29.3)] on darwin
+   Type "help", "copyright", "credits" or "license" for more information.
+   
+We have entered into Python console from the shell and we can
```
Comment (wjones127) [inline docs/source/developers/guide/tutorials/python_tutorial.rst:182]:
```suggestion
We have entered into the Python console from the shell and we can
```

File: docs/source/developers/guide/tutorials/python_tutorial.rst [hunk 11]
```diff
@@ -25,3 +25,532 @@
...
+   ...      pa.field('min-', pa.int64()),
+   ...      pa.field('max+', pa.int64()),
+   ...    ])
+   >>> pa.scalar([('min-', 3), ('max+', 9)], type=ty)
+   <pyarrow.StructScalar: [('min-', 3), ('max+', 9)]>
+
+.. note::
+
+   In cases where we don't yet have good documentation, unit tests
+   can be a good place to look for code examples
```
Comment (wjones127) [inline docs/source/developers/guide/tutorials/python_tutorial.rst:260]:
```suggestion
   can be a good place to look for code examples.
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
Just did a second pass and have a few more minor suggestions. Getting close to ready IMO 😄

File: docs/source/developers/guide/tutorials/python_tutorial.rst (hunk 9 above)
Comment (AlenkaF) [reply docs/source/developers/guide/tutorials/python_tutorial.rst:167]: Oops, thanks!

**Review by wjones127 — APPROVED** [review]
Awesome! Looks great to me.

Maybe @thisisnic will want to look at it too.

**Comment by thisisnic** [comment]
> Awesome! Looks great to me.
> 
> Maybe @thisisnic will want to look at it too.

Given you've made quite a few suggestions on this PR already @wjones127 , unless there's anything specific you'd like me to look at, then if you and @AlenkaF are happy with it, I have every faith it's all good!

**Pull request closed without merging**